render preview image url in card, when imageUrl is provided

bikrantjajware opened the pull request.

rickstaa commented on src/common/Card.js:
```diff
@@ -199,6 +209,18 @@ class Card {
     `;
   };
 
+  renderImage = () => {
```
Please add a docstring.

rickstaa commented on src/common/Card.js:
```diff
@@ -36,6 +40,8 @@ class Card {
 
+    this.imageHeight = imageHeight;
+    this.imageUrl = imageUrl;
     // returns theme based colors with proper overrides and defaults
```
```suggestion

    // returns theme based colors with proper overrides and defaults
```

rickstaa commented on src/cards/repo-card.js:
```diff
@@ -61,6 +61,7 @@ const renderRepoCard = (repo, options = {}) => {
     isTemplate,
     starCount,
     forkCount,
+    openGraphImageUrl,
```
I think you should add this as an optional option to the pin card:

https://github.com/anuraghazra/github-readme-stats/blob/ff169849880e3c2b0d89cc8d95b6542e71fc1906/api/pin.js#L13-L27

rickstaa reviewed (changes requested):
Absolutely:

Hey @bikrantjajware, kudos on your pull request. After a thorough review, I've tagged it with `hacktoberfest-accepted` as it aligns with the specified criteria. To proceed with merging into the main branch, we'll need:

- [ ] An optional query parameter facilitating user-enabled image display.
- [ ] Comprehensive documentation outlining the new feature.
- [ ] Test cases validating the functionality of the introduced feature.

Your attention to these details will help smooth the integration process. Cheers!

bikrantjajware commented:
Hello @rickstaa ,
thanks for the suggestions. Addressed the comments and added tests + update documentations. Let me know if i should cover any more test scenarios :)

rickstaa reviewed (approved):
I don't know about the header formatting since, for some repositories in which a custom social image is set, the styling doesn't look right. We could add a small top margin, but then again, people are likely pinning their repositories and can fix the image header size if they want. @qwerty, what do you think?

[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=anuraghazra&repo=github-readme-stats&show_image=true)](https://github.com/anuraghazra/github-readme-stats)

```md
[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=anuraghazra&repo=github-readme-stats&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=github-emoji-picker&show_image=true)](https://github.com/anuraghazra/github-readme-stats)

```md
[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=github-emoji-picker&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=tmux-notify&show_image=true)](https://github.com/anuraghazra/github-readme-stats)

```md
[![Readme Card](https://github-readme-stats-git-test-card-image-header-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=tmux-notify&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

rickstaa reviewed (changes requested):
I'm not sure what is happening, but it looks like when deployed, the images don't show up when they are used in markdown, while they show up when the API link is visited directly. Maybe it's a caching problem?

![image](https://github.com/anuraghazra/github-readme-stats/assets/17570430/e89ec1f0-6161-4b01-977f-ebf3b01afac0)

rickstaa commented:
> Hello @rickstaa , thanks for the suggestions. Addressed the comments and added tests + update documentations. Let me know if i should cover any more test scenarios :)

Thanks for addressing my points. I think your pull request is good to be merged. I however noticed that it doesn't seem to work on the deployed instance 🤔.

bikrantjajware commented:
Looks like github hides the image behind some proxy (https://github.com/orgs/community/discussions/30326), and probably the proxy service (camo) is unable to read the repo image (maybe timed out error). I'll look into possible solutions
Maybe change in the way we render image in svg

bikrantjajware commented:
Hello @rickstaa , i'm converting the repo_image to a base64 string data and using that to render the header image. Tested with deployed vercel url probably should work now.

rickstaa commented:
Thanks a bunch for incorporating the requested tweaks and squashing that bug! The changes look stellar.

```md
[![GitHub Readme Card](https://github-readme-stats-wlrp4x0wa-rickstaa.vercel.app/api/pin/?username=anuraghazra&repo=github-readme-stats&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

And the cleanup on the second snippet is on point.

```md
[![GitHub Readme Card](https://github-readme-stats-wlrp4x0wa-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=tmux-notify&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

With that docstring in place, your code gets a solid thumbs up. The only puzzle piece left is nailing down the styling for repositories rocking custom social images (check out the example below). I've tossed the topic over to @qwerty541 for a second opinion 👍🏻.

```md
[![GitHub Readme Card](https://github-readme-stats-wlrp4x0wa-rickstaa.vercel.app/api/pin/?username=rickstaa&repo=github-emoji-picker&show_image=true)](https://github.com/anuraghazra/github-readme-stats)
```

bikrantjajware commented:
Thanks @rickstaa  :) ,  for custom images  we can set social images in repo > settings > social preview and upload an image

rickstaa commented:
> Thanks @rickstaa :) , for custom images we can set social images in repo > settings > social preview and upload an image

Hello @bikrantjajware, thank you for your prompt response. I apologize for any confusion in my previous message. While I am familiar with setting a custom image (as demonstrated in example three), my current query pertains to the card styling when users opt for such custom images. Specifically, I've observed a lack of padding at the top. Could you kindly provide clarification on this matter? Your insights would be greatly appreciated. Thank you.

rickstaa commented:
> Hello @rickstaa , do you mean something like this with padding-top, it looks a bit standing out <img alt="image" width="403" src="https://user-images.githubusercontent.com/34072248/281049543-672bd3ce-cc39-4b50-8702-b40ba969e50d.png">
> 
> or do you mean like filling the entire card width <img alt="image" width="399" src="https://user-images.githubusercontent.com/34072248/281050255-1f77f745-896f-49c1-bde3-539008c80037.png">

I meant the first option since the second option would be undoable, given that users can define any social image they like 😅. Let's see which formatting @qwerty541 prefers 👍🏻.

bikrantjajware commented:
oh i see, probably customizing with url params ? 
for 2nd option we can  set `preserveAspectRatio` to `none` on image tag, though it stretches to fill entire width :)

rickstaa commented:
> oh, i see, probably customizing with url params ? for 2nd option we can set `preserveAspectRatio` to `none` on image tag, though it stretches to fill entire width :)

@bikrantjajware, thanks for the fast response. We don't need to add a URL parameter for the top padding. We simply have to decide on what styling we want to apply. The options we now discussed are:

- Use top padding.
- Se  `preserveAspectRatio` to `none`.

Let's wait and see what the other collaborators think about this 👍🏻. Also, please bear in mind that since we are a small team, PRs are merged in the main branch based on the community's interest (see https://github.com/anuraghazra/github-readme-stats/issues/1935). After that, we typically merge popular pull requests when more than 2 collaborators approve them. It could, therefore, take some time until your PR shows up on the main branch 😅.
